### This Python code does the part of data preparation for analazing the NBA results of any season


Primer cal importar les dues llibreries de Python que caldrà utlilitzar per la preparació de les dades ja extretes que volem fer.

Les dues llibreries són:

-Numpy: És un package de Python fonamental en la programació cientifica amb Pyhton. Conté a part de moltes altres coses, un potent objecte (N-dimensional array), funcions sofisticades, eines per integrar en codis de C/C++ i Fortran, funcion d'algebra lineal, transformades de Fourier i funcions de creació de numero aleatoris.

A part del seu ús per programació cientifica, NumPy es pot utilitzar com una eina eficient per guardar qualsevol tipus de dades. Això fa que NumPy pugui treballar de manera ràpida i eficient amb una gran varietat de bases de dades. Per aquest projecte serà la funció principal que li donarem a NumPy ja que ens permetrà guardar les dades de forma eficient i també ens permetrà poder manipular aquestes dades de forma fàcil.
**Posar exemple array i explicar millor que és**

-Pandas: Es tracta d'una llibreria de Python destinada bàsicament a l'anàlisi de dades, ja que proporciona unes estructures de dades flexibles i que permeten treballar amb les dades de forma molt eficient. Ofereix les següents estructures de dades:


Series: Són arrays unidimensionals amb indexació (arrays amb índex) i són similars als diccionaris, de fet les series poden generar-se a través de diccionaris o de llistes.
**Posar exemple serie**
     
DataFrame: Són estructures de dades similars a les taules de bases de dades realacionals com s'utilitzen en el extensament conegut llenguatges SQL.
**Posar exemple dataframe**
     
Panel, Panel4D y PanelND: Aquestes estructures permeten treballar amb dades de més de dos dimensions però és algo complex i no s'utilitzarà per a aquest projecte.



In [1]:
#importing required libraries

import pandas as pd
import numpy as np


#### Important les dades que volem analitzar

En la cel·la que hi ha  continuació importarem els fitxers en format **.csv** que hem obtingut en la recopilació de dades per el projecte.
Els fitxer en format **.csv** (de l'anglès comma-separated values) són un tipus de document en format obert senzill per representar dades en forma de taula, en què les columnes se separen per comes (o punt i coma on la coma és el separador decimal: Catalunya, França, Itàlia...) i les files per salts de línia. Els camps que continguin una coma, un salt de línia o una cometa doble han de ser tancats entre cometes dobles.

El format CSV és molt senzill i no indica un joc de caràcters concret, ni com van situats els bytes, ni el format pel salt de línia. Aquests punts s'han d'indicar molts cops en obrir el fitxer, per exemple, amb un full de càlcul. 

Per tal de importar la taula que contenen els fitxers **.csv** ho farem amb la funció de Python Pandas pandas.read_csv("Carpeta_on_es_troba/nom_del_fitxer"). Aquesta funció ens convertirà el fitxer .csv en un un dataframe de la llibreria pandas que ja hem explicat anteriorment.

El primer fitxer que importarem conté la següent informació sobre cada partit jugat durant la temporada 19/20 de la NBA a cada columna.

- **Date**: Conté un string amb la data del partit amb format (dia, mes dia, any).
- **Start(ET)**: hora en què va començar el partit en hora de la costa oest d'Estats Units.
- **Visitor**: Nom complet de l'equip visitant en format string.
- **PTSV**:Punts anotats per l'equip visitant al final del partit en format int.
- **Home**: Nom complet de l'equip local en format string.
- **PTSH**:Punts anotats per l'equip local al final del partit en format int.
- **Overtime**: Indica amb un string 'OT' si hi ha hagut pròrroga en el partit, si no hi ha hagut pròrroga ho indica amb una casella en blanc.
- **Attend**: Indica l'assistència de públic en format string.

A continuació importem el fitxer i en veiem unes quantes files del mateix.

In [2]:
data = pd.read_csv("NBA_project/NBA_19_20_results.csv")
data

Date Start (ET)               Visitor  PTSV  \
0    Tue, Oct 22, 2019      8:00p  New Orleans Pelicans   122   
1    Tue, Oct 22, 2019     10:30p    Los Angeles Lakers   102   
2    Wed, Oct 23, 2019      7:00p         Chicago Bulls   125   
3    Wed, Oct 23, 2019      7:00p       Detroit Pistons   119   
4    Wed, Oct 23, 2019      7:00p   Cleveland Cavaliers    85   
..                 ...        ...                   ...   ...   
966  Tue, Mar 10, 2020     10:30p         Brooklyn Nets   104   
967  Wed, Mar 11, 2020      7:00p       Detroit Pistons   106   
968  Wed, Mar 11, 2020      7:30p       New York Knicks   136   
969  Wed, Mar 11, 2020      7:30p     Charlotte Hornets   109   
970  Wed, Mar 11, 2020      8:00p        Denver Nuggets    97   

                     Home  PTSH Overtime Attend.  
0         Toronto Raptors   130       OT  20,787  
1    Los Angeles Clippers   112      NaN  19,068  
2       Charlotte Hornets   126      NaN  15,424  
3          Indiana Pacers   110      NaN  17,923  
4           Orlando Magic    94      NaN  18,846  
..                    ...   ...      ...     ...  
966    Los Angeles Lakers   102      NaN  18,997  
967    Philadelphia 76ers   124      NaN  20,172  
968         Atlanta Hawks   131       OT  15,393  
969            Miami Heat    98      NaN    19,6  
970      Dallas Mavericks   113      NaN  20,302  

[971 rows x 8 columns]

El següent fitxer de dades que importarem conté una gran quantitat d'estadística avançada de tots els jugadors de la NBA durant aquesta temporada. Malgrat això només ens interesserà per el projecte tres columnes que són les següents.

- **Player**: conté el nom complet del jugador en format string.
- **Tm**: conté el nom de l'equip en què juga abreviat amb 3 lletres majuscules, també en format string.
- **STAR**: conté un 1 si el jugador és considerat una estrella de la lliga per el rànking de top 50 jugadors de la NBA que es fa cada temporada. En cas que el jugador no estigui en aquest Top 50 la taula contindrà un 0.
https://www.interbasket.net/news/2019/list-espn-top-100-best-nba-players-2019-20-season-nbarank

A continuació importem el fitxer i en veiem unes quantes files del mateix.

In [3]:
data_players=pd.read_csv("NBA_project/players_stats_19_20.csv")
data_players

Player Pos  Age   Tm   G    MP   PER    TS%   3PAr    FTr  \
0             James Harden  SG   30  HOU  61  2241  28.4  0.616  0.555  0.519   
1    Giannis Antetokounmpo  PF   25  MIL  57  1763  31.6  0.608  0.238  0.500   
2             LeBron James  PG   35  LAL  60  2094  26.0  0.582  0.324  0.292   
3             Nikola Jokic   C   24  DEN  65  2101  25.0  0.604  0.233  0.277   
4            Anthony Davis  PF   26  LAL  55  1889  28.2  0.614  0.195  0.460   
..                     ...  ..  ...  ...  ..   ...   ...    ...    ...    ...   
494             RJ Barrett  SG   19  NYK  56  1704  10.7  0.479  0.271  0.349   
495          Dillon Brooks  SG   24  MEM  65  1851  11.1  0.508  0.387  0.192   
496        De'Andre Hunter  SF   22  ATL  63  2018   8.6  0.521  0.445  0.211   
497           Jordan Poole  SG   20  GSW  57  1274   7.2  0.454  0.528  0.237   
498         Darius Garland  PG   20  CLE  59  1824   8.5  0.498  0.423  0.103   

     ...  USG%  OWS  DWS    WS   WS/48  OBPM  DBPM   BPM  VORP  STAR  
0    ...  36.4  8.7  2.7  11.5   0.245   7.9   1.2   9.0   6.3     1  
1    ...  37.4  5.6  4.8  10.4   0.282   7.4   4.1  11.5   6.0     1  
2    ...  31.6  6.1  3.4   9.5   0.218   6.8   1.9   8.7   5.7     1  
3    ...  26.6  6.0  3.2   9.2   0.209   5.4   2.3   7.6   5.1     1  
4    ...  29.7  6.2  4.1  10.3   0.262   5.8   2.8   8.5   5.0     1  
..   ...   ...  ...  ...   ...     ...   ...   ...   ...   ...   ...  
494  ...  24.0 -1.6  1.1  -0.5 -15.000  -2.8  -1.5  -4.3  -1.0     0  
495  ...  25.0 -0.3  1.3   1.0   0.025  -2.8  -1.5  -4.3  -1.1     0  
496  ...  17.5 -0.4  0.5   0.1   0.001  -2.8  -1.8  -4.7  -1.4     0  
497  ...  21.1 -1.6  0.4  -1.2 -47.000  -4.4  -2.2  -6.6  -1.5     0  
498  ...  20.7 -1.3  0.0  -1.3 -35.000  -2.7  -2.9  -5.6  -1.7     0  

[499 rows x 27 columns]

El següent fitxer conté el preu mitjà de les apostes que hi havia per a cada partit abans que el mateix comencés. L'aposta és el preu per euro apostat per la victòria de l'equip en questió. Per exemple si l'aposta és 2.5 vol dir que si s'aposta 1 euro en cas que l'equip guanyi en guanyaràs 2.5. Les columnes en questió són:

- **NEW DATE**: conté la data del partit en format (dd/mm/aa) en string.
- **HOME**: conté un string amb el nom de l'equip local.
- **VISITOR**: conté un string amb el nom de l'equip visitant.
- **BET HOME**: conté el valor de l'aposta en format float per l'equip local.
- **BET VISITOR**: conté el valor de l'aposta en format float per l'equip visitant.

A continuació importem el fitxer i en veiem unes quantes files del mateix.

In [4]:
data_bets=pd.read_csv("NBA_project/new_bets_19_20.csv")
data_bets

NEW DATE                    HOME                VISITOR  BET HOME  \
0    11/03/20        Dallas Mavericks         Denver Nuggets      2.21   
1    11/03/20          Atlanta Hawks         New York Knicks      1.48   
2    11/03/20             Miami Heat       Charlotte Hornets      1.15   
3    11/03/20      Philadelphia 76ers        Detroit Pistons      1.12   
4    10/03/20  Golden State Warriors    Los Angeles Clippers      7.46   
..        ...                     ...                    ...       ...   
966  23/10/19       Charlotte Hornets          Chicago Bulls      2.39   
967  23/10/19         Indiana Pacers         Detroit Pistons      1.33   
968  23/10/19           Orlando Magic    Cleveland Cavaliers      1.20   
969  22/10/19    Los Angeles Clippers     Los Angeles Lakers      2.39   
970  22/10/19         Toronto Raptors   New Orleans Pelicans      1.36   

     BET VISITOR  
0           1.71  
1           2.72  
2           5.88  
3           6.82  
4           1.10  
..           ...  
966         1.62  
967         3.45  
968         4.91  
969         1.61  
970         3.28  

[971 rows x 5 columns]

El següent fitxer conté informació de la distància entre totes les ciutats que tenen equips de la NBA en kilòmetres. Conté les següents columnes de dades:

- **City Origin**: Conté un string amb el nom de l'equip de la ciutat origen. Per exemple 'Washington Wizards' es refereix a la ciutat de Washington DC dels EEUU.
- **City Destiny**: Conté un string amb el nom de l'equip de la ciutat destí del viatge. Per exemple 'Denver Nuggets' es refereix a la ciutat de Denver dels EEUU.
- **Distance (km)**: conté la distància en format float dels km en línia recta entre les dues ciutats on es farà el viatge. Aquesta taula s'ha aconseguit utilitzant un altre codi fent web scrapping que ja s'ha explicat anteriorment.

In [5]:
data_distance=pd.read_csv("NBA_project/distances_NBA.csv")
data_distance

City Origin            City Destiny  Distance (Km)
0      Washington Wizards          Denver Nuggets         2395.0
1      Washington Wizards  Minnesota Timberwolves         1503.0
2      Washington Wizards   Oklahoma City Thunder         1855.0
3      Washington Wizards  Portland Trail Blazers         3786.0
4      Washington Wizards               Utah Jazz         2976.0
..                    ...                     ...            ...
872    Los Angeles Lakers            Phoenix Suns          587.5
873    Los Angeles Lakers           Brooklyn Nets         3966.0
874  Los Angeles Clippers      Los Angeles Lakers            0.0
875    Los Angeles Lakers    Los Angeles Clippers            0.0
876        Denver Nuggets          Denver Nuggets            0.0

[877 rows x 3 columns]

##### Petit anàlisi del dataframe 'data'

Utilitzant la funció de Pandas describe( ), obtenim la següent informació estadística bàsica de totes les columnes en format numèric del dataframe en aquest cas PTSV, PTSH.

In [6]:
data.describe()

PTSV        PTSH
count  971.000000  971.000000
mean   110.359423  112.533471
std     12.147783   12.578728
min     76.000000   73.000000
25%    102.000000  104.000000
50%    110.000000  112.000000
75%    119.000000  120.000000
max    159.000000  158.000000

Podem observar clarament que el nombre de partits del dataframe és 971, també podem observar que la mitjana de punts anotats per els equips que juguen a casa és superior a la dels equips que juguen a fora de casa. A més també podem veure els valors màxims i minims d'anotació i el valor dels quartils.

Per tal de preparar les dades per procedir a l'anàlisi primer de tot caldrà manipular aquestes dades de forma que ens puguin ser útils i extreuren tota la informació possible per obtenir la màxima precisió en les previsions.

Primer de tot eliminem amb la funció drop( ) de Pandas les columnes del dataframe data que no ens interessen. En aquest cas es tracta de l'hora del partit que no s'ha considerat que sigui rellevant a l'hora de poder preveure el resultat.

In [7]:
# Borrem la columna que conté la hora del partit ja que aquesta no ens dóna cap informació rellevant

data = data.drop('Start (ET)', axis=1)

A continuació també cal arreglar el format de la columna Overtime, substituirem els valors NaN per un 0 utilitzant la funció de pandas data.fillna(0). També canviarem els strings que contenen 'OT' o '2OT', '3OT', '4OT', etc. Per un 1. considerem que tant si hi ha hagut una pròrroga com quatre el mateix cas per tal de simplificar la columna a només 0 o 1.

In [8]:
#Substituim 'NaN' a la columna Overtime per un 0 i'OT' per 1

data=data.fillna(0)
data=data.replace(['OT','2OT','3OT','4OT'],1)
data

Date               Visitor  PTSV                  Home  \
0    Tue, Oct 22, 2019  New Orleans Pelicans   122       Toronto Raptors   
1    Tue, Oct 22, 2019    Los Angeles Lakers   102  Los Angeles Clippers   
2    Wed, Oct 23, 2019         Chicago Bulls   125     Charlotte Hornets   
3    Wed, Oct 23, 2019       Detroit Pistons   119        Indiana Pacers   
4    Wed, Oct 23, 2019   Cleveland Cavaliers    85         Orlando Magic   
..                 ...                   ...   ...                   ...   
966  Tue, Mar 10, 2020         Brooklyn Nets   104    Los Angeles Lakers   
967  Wed, Mar 11, 2020       Detroit Pistons   106    Philadelphia 76ers   
968  Wed, Mar 11, 2020       New York Knicks   136         Atlanta Hawks   
969  Wed, Mar 11, 2020     Charlotte Hornets   109            Miami Heat   
970  Wed, Mar 11, 2020        Denver Nuggets    97      Dallas Mavericks   

     PTSH  Overtime Attend.  
0     130         1  20,787  
1     112         0  19,068  
2     126         0  15,424  
3     110         0  17,923  
4      94         0  18,846  
..    ...       ...     ...  
966   102         0  18,997  
967   124         0  20,172  
968   131         1  15,393  
969    98         0    19,6  
970   113         0  20,302  

[971 rows x 7 columns]

A continuació volem crear una funció que modifiqui el format de la data del fitxer data a un format igual que el que tenim en el fitxer data_bets d'aquesta manera ens serà molt més fàcil manipular i creuar dades dels diferents dataframes. La funció canvia el format la data tal i com es veu en l'exemple següent (p.e. 'Tue, Oct 22, 2019' to '22/10/19')


In [9]:

def new_date_format(data_2change,dim):
    d=[]
    month=0
    date_month=' '
    # El diccionari només inclou els mesos en què es juga la temporada NBA
    months=[['Oct','10'],['Nov','11'],['Dec','12'],['Jan','01'],['Feb','02'],['Mar','03'],['Apr','04']]
    for i in range (dim):
        date=data_2change[i].split(', ')
        date_month_day=date[1].split(' ')
        for e in months:
            if date_month_day[0]==e[0]:
                month=e[1]
        if int(date_month_day[1])<10:
            d=np.append(d,['0'+date_month_day[1]+'/'+month+'/'+date[2][2::]])
        else:
            d=np.append(d,[date_month_day[1]+'/'+month+'/'+date[2][2::]])

    return d

# Afegim la columna 'New Date' al dataframe data i eliminenm la data en format antic 'Date'.
new_date= new_date_format(data['Date'],len(data['Date']))
data = data.drop('Date', axis=1)
data['New Date']=new_date
data

Visitor  PTSV                  Home  PTSH  Overtime Attend.  \
0    New Orleans Pelicans   122       Toronto Raptors   130         1  20,787   
1      Los Angeles Lakers   102  Los Angeles Clippers   112         0  19,068   
2           Chicago Bulls   125     Charlotte Hornets   126         0  15,424   
3         Detroit Pistons   119        Indiana Pacers   110         0  17,923   
4     Cleveland Cavaliers    85         Orlando Magic    94         0  18,846   
..                    ...   ...                   ...   ...       ...     ...   
966         Brooklyn Nets   104    Los Angeles Lakers   102         0  18,997   
967       Detroit Pistons   106    Philadelphia 76ers   124         0  20,172   
968       New York Knicks   136         Atlanta Hawks   131         1  15,393   
969     Charlotte Hornets   109            Miami Heat    98         0    19,6   
970        Denver Nuggets    97      Dallas Mavericks   113         0  20,302   

     New Date  
0    22/10/19  
1    22/10/19  
2    23/10/19  
3    23/10/19  
4    23/10/19  
..        ...  
966  10/03/20  
967  11/03/20  
968  11/03/20  
969  11/03/20  
970  11/03/20  

[971 rows x 7 columns]

Per tal de poder predir de forma correcte els partits del principi de temporada quan encara no hi ha suficient volum de dades per tal que el model pugui aprendre, utilitzarem la previsió de victòries per cada equip que fa las Vegas al principi de la temporada. És a dir el nombre de victòries que s'esperen que un equip aconseguirà al final de la temporada sobre un total de 82 partits que disputa cada equip. Guardarem aquestes dades en un diccionari que anomenarem Vegas_pred. 

Com es pot veure conté com a clau el nom coomplet de l'equip i com a valor associat un valor float amb el numero de victories esperades sobre un total de 82 possibles.
https://www.basketball-reference.com/leagues/NBA_2020_preseason_odds.html

In [10]:
Vegas_pred={'Toronto Raptors':45, 'Boston Celtics':49.5, 'Philadelphia 76ers':54.5, 'Brooklyn Nets':47, 'New York Knicks':27,'Milwaukee Bucks':57, 'Indiana Pacers':48.5, 'Chicago Bulls':30.5,'Detroit Pistons':37.5, 'Cleveland Cavaliers':24,'Miami Heat':43.5, 'Orlando Magic':40.5,'Washington Wizards':28.5,'Charlotte Hornets':24,'Atlanta Hawks':36,'Denver Nuggets':52,'Utah Jazz':52.5,'Oklahoma City Thunder':31,'Portland Trail Blazers':47.5, 'Minnesota Timberwolves':35,'Los Angeles Lakers':51.5,'Los Angeles Clippers':55.5,'Sacramento Kings':37,'Phoenix Suns':27,'Golden State Warriors':47,'Houston Rockets':52,'Dallas Mavericks':41,'Memphis Grizzlies':25.5,'New Orleans Pelicans':39,'San Antonio Spurs':43.5}
Vegas_pred

{'Toronto Raptors': 45,
 'Boston Celtics': 49.5,
 'Philadelphia 76ers': 54.5,
 'Brooklyn Nets': 47,
 'New York Knicks': 27,
 'Milwaukee Bucks': 57,
 'Indiana Pacers': 48.5,
 'Chicago Bulls': 30.5,
 'Detroit Pistons': 37.5,
 'Cleveland Cavaliers': 24,
 'Miami Heat': 43.5,
 'Orlando Magic': 40.5,
 'Washington Wizards': 28.5,
 'Charlotte Hornets': 24,
 'Atlanta Hawks': 36,
 'Denver Nuggets': 52,
 'Utah Jazz': 52.5,
 'Oklahoma City Thunder': 31,
 'Portland Trail Blazers': 47.5,
 'Minnesota Timberwolves': 35,
 'Los Angeles Lakers': 51.5,
 'Los Angeles Clippers': 55.5,
 'Sacramento Kings': 37,
 'Phoenix Suns': 27,
 'Golden State Warriors': 47,
 'Houston Rockets': 52,
 'Dallas Mavericks': 41,
 'Memphis Grizzlies': 25.5,
 'New Orleans Pelicans': 39,
 'San Antonio Spurs': 43.5}

La següent funció ens permetrà extreure el primer paràmetre que ens servirà per fer la predicció de resultats. aquest paràmetre serà la diferència de % de victòries entre l'equip local i l'equip visitant just abans de començar el partit en questió.

Primer de tot cal explicar en què consisteix el percentatge de victòries abans de cada partit. Per posar un exemple si un equip porta 15 partits jugats i n'ha guanyat  7 el seu percentatge de victòries en aquell moment serà del 46.6% i si per exemple llavors guanya 5 partits seguits en portarà guanyats 12 sobre 20 lo que representa un 60%. Aquest percentatge de victòries actualitzat ens permet monitoritzar el moment de forma de cada equip  i el global de la temporada fins aquell moment donant-nos una idea molt més acurada de les possibilitats reals de l'equip de guanyar cada partit en concret.

Per tal de obtenir aquest percentatge de victòries actualitzar per cada partit i equip haurem de crear els següent diccionari anomenat w_percent.
També crearem dos arrays buits on emmagtzamarem els percentatges de victòries actualitzats i els afegirem al dataframe.

Aquest diccionari contindrà com a clau el nom de l'equip de l'NBA en questió i com a valor associat a la clau  una llista amb dos integers el primer serà el nombre de victòries obtingudes fins el moment i el segon el nombre total de partits disputats fins aquell moment.

No obstant duran els primers 120 partits de la temporada ulitzarem el percentatge de victòries previst per les Vegas al principi de temporada que hem guardat al diccionari Vegas_pred, d'aquesta manera evitem errors en els primers partits per manca de dades. un cop disputats 120 partits de la temporada cada equip n'haurà disputat com a minim 8 i per tant el percentatge de victories obtingut ja començarà a ser més rellevant (120 partits és un nombre arbitràri que pot ser modificat).

La funció transcorre de la següent manera:

1. Creem un loop amb for i in range( ) de la longitud del dataframe data d'aquesta manera analitzem cada fila que representa un partit. 

2. Apliquem un if per veure si ha guanyat l'equip local o el visitant i dins d'aquest if un altre per veure si estem dins dels primers 120 partits o no. 

3. Un cop analitzades aquestes dues condicions primer de tot consultarem el valor actual del diccionari que conté el percentatge de victòries per cada equip local i visitant i l'afegirem a la array corresponent que hem creat anteriorment. Finalment modificarem el valor del diccionari de cada equip afegint un +1 a partits disputats i un +1 en cas que l'equip hagi guanyat.

4. la funció retornarà el diccionari amb el percentatge de victòries de cada equip i dos arrays que contenen els percentatges de victòries per cada partit dels equips locals i visitants.

Finalment restem els dos arrays per tal de obtenir un tercer array que ens doni la diferencia entre els dos percentatges de victories.





In [11]:
def win_percent(data, dim):
    w_percent={}
    w_percent_visitor=[]
    w_percent_local=[]
    w_percent_global=[]
    for i in range (dim):
        if data['Visitor'][i] not in w_percent:
            w_percent[data['Visitor'][i]]=[0,0]
           
        if data['Home'][i] not in w_percent:
            w_percent[data['Home'][i]]=[0,0]   
            

        if int(data['PTSV'][i])>int(data['PTSH'][i]):
            if i<120:
                w_percent_visitor=np.append(w_percent_visitor, "{0:.3f}".format(float(Vegas_pred[data['Visitor'][i]])/float(82)))
                w_percent_local=np.append(w_percent_local, "{0:.3f}".format(float(Vegas_pred[data['Home'][i]])/float(82)))
            else:
                w_percent_visitor=np.append(w_percent_visitor, "{0:.3f}".format(float(w_percent[data['Visitor'][i]][0])/float(w_percent[data['Visitor'][i]][1])))
                w_percent_local=np.append(w_percent_local, "{0:.3f}".format(float(w_percent[data['Home'][i]][0])/float(w_percent[data['Home'][i]][1])))
            w_percent[data['Visitor'][i]][0]= w_percent[data['Visitor'][i]][0]+1
            w_percent[data['Visitor'][i]][1]= w_percent[data['Visitor'][i]][1]+1
            w_percent[data['Home'][i]][1]=w_percent[data['Home'][i]][1]+1
            
        else:
            if i<120:
                w_percent_visitor=np.append(w_percent_visitor, "{0:.3f}".format(float(Vegas_pred[data['Visitor'][i]])/float(82)))
                w_percent_local=np.append(w_percent_local, "{0:.3f}".format(float(Vegas_pred[data['Home'][i]])/float(82)))
            else:
                w_percent_visitor=np.append(w_percent_visitor,"{0:.3f}".format(float(w_percent[data['Visitor'][i]][0])/float(w_percent[data['Visitor'][i]][1])))
                w_percent_local=np.append(w_percent_local, "{0:.3f}".format(float(w_percent[data['Home'][i]][0])/float(w_percent[data['Home'][i]][1])))
            w_percent[data['Home'][i]][0]= w_percent[data['Home'][i]][0]+1
            w_percent[data['Home'][i]][1]= w_percent[data['Home'][i]][1]+1
            w_percent[data['Visitor'][i]][1]=w_percent[data['Visitor'][i]][1]+1


    return w_percent,w_percent_local,w_percent_visitor

win_percentage,w_percent_l,w_percent_v=win_percent(data,len(data))   

#print (w_percent_v)
#print (w_percent_l)
w_percent=[]
for i in range(len(w_percent_l)):
    w_percent=np.append(w_percent,"{0:.3f}".format(float(w_percent_l[i])-float(w_percent_v[i])))

print(win_percentage)
data['%W']=w_percent#this gets the difference of the win percentage of the two teams that play
data

{'New Orleans Pelicans': [28, 64], 'Toronto Raptors': [46, 64], 'Los Angeles Lakers': [49, 63], 'Los Angeles Clippers': [44, 64], 'Chicago Bulls': [22, 65], 'Charlotte Hornets': [23, 65], 'Detroit Pistons': [20, 66], 'Indiana Pacers': [39, 65], 'Cleveland Cavaliers': [19, 65], 'Orlando Magic': [30, 65], 'Minnesota Timberwolves': [19, 64], 'Brooklyn Nets': [30, 64], 'Memphis Grizzlies': [32, 65], 'Miami Heat': [41, 65], 'Boston Celtics': [43, 64], 'Philadelphia 76ers': [39, 65], 'Washington Wizards': [24, 64], 'Dallas Mavericks': [40, 67], 'New York Knicks': [21, 66], 'San Antonio Spurs': [27, 63], 'Oklahoma City Thunder': [40, 64], 'Utah Jazz': [41, 64], 'Sacramento Kings': [28, 64], 'Phoenix Suns': [26, 65], 'Denver Nuggets': [43, 65], 'Portland Trail Blazers': [29, 66], 'Atlanta Hawks': [20, 67], 'Milwaukee Bucks': [53, 65], 'Houston Rockets': [40, 64], 'Golden State Warriors': [15, 65]}


Visitor  PTSV                  Home  PTSH  Overtime Attend.  \
0    New Orleans Pelicans   122       Toronto Raptors   130         1  20,787   
1      Los Angeles Lakers   102  Los Angeles Clippers   112         0  19,068   
2           Chicago Bulls   125     Charlotte Hornets   126         0  15,424   
3         Detroit Pistons   119        Indiana Pacers   110         0  17,923   
4     Cleveland Cavaliers    85         Orlando Magic    94         0  18,846   
..                    ...   ...                   ...   ...       ...     ...   
966         Brooklyn Nets   104    Los Angeles Lakers   102         0  18,997   
967       Detroit Pistons   106    Philadelphia 76ers   124         0  20,172   
968       New York Knicks   136         Atlanta Hawks   131         1  15,393   
969     Charlotte Hornets   109            Miami Heat    98         0    19,6   
970        Denver Nuggets    97      Dallas Mavericks   113         0  20,302   

     New Date      %W  
0    22/10/19   0.073  
1    22/10/19   0.049  
2    23/10/19  -0.079  
3    23/10/19   0.134  
4    23/10/19   0.201  
..        ...     ...  
966  10/03/20   0.330  
967  11/03/20   0.286  
968  11/03/20  -0.005  
969  11/03/20   0.297  
970  11/03/20  -0.081  

[971 rows x 8 columns]

#### This function gets the %W of every team as local or as visitor at the current game

In [12]:
def w_perc_local_visitor(data, dim):
    w_percent_as_local={}
    w_percent_as_visitor={}
    w_percent_visitor=[]
    w_percent_local=[]
    for i in range (dim):
        if data['Visitor'][i] not in w_percent_as_visitor:
            w_percent_as_visitor[data['Visitor'][i]]=[0,0]
           
        if data['Home'][i] not in w_percent_as_local:
            w_percent_as_local[data['Home'][i]]=[0,0] 

        if int(data['PTSV'][i])>int(data['PTSH'][i]):
            if i<120:
                w_percent_visitor=np.append(w_percent_visitor, "{0:.3f}".format(float(0.45*Vegas_pred[data['Visitor'][i]])/float(41)))
                w_percent_local=np.append(w_percent_local, "{0:.3f}".format(float(0.55*Vegas_pred[data['Home'][i]])/float(41)))

            else:
                w_percent_visitor=np.append(w_percent_visitor, "{0:.3f}".format(float(w_percent_as_visitor[data['Visitor'][i]][0])/float(w_percent_as_visitor[data['Visitor'][i]][1])))
                w_percent_local=np.append(w_percent_local, "{0:.3f}".format(float(w_percent_as_local[data['Home'][i]][0])/float(w_percent_as_local[data['Home'][i]][1])))
            w_percent_as_visitor[data['Visitor'][i]][0]= w_percent_as_visitor[data['Visitor'][i]][0]+1
            w_percent_as_visitor[data['Visitor'][i]][1]= w_percent_as_visitor[data['Visitor'][i]][1]+1
            w_percent_as_local[data['Home'][i]][1]=w_percent_as_local[data['Home'][i]][1]+1

        else:

            if i<120:

                w_percent_visitor=np.append(w_percent_visitor, "{0:.3f}".format(float(0.45*Vegas_pred[data['Visitor'][i]])/float(41)))
                w_percent_local=np.append(w_percent_local, "{0:.3f}".format(float(0.55*Vegas_pred[data['Home'][i]])/float(41)))
            else:
                w_percent_visitor=np.append(w_percent_visitor,"{0:.3f}".format(float(w_percent_as_visitor[data['Visitor'][i]][0])/float(w_percent_as_visitor[data['Visitor'][i]][1])))
                w_percent_local=np.append(w_percent_local, "{0:.3f}".format(float(w_percent_as_local[data['Home'][i]][0])/float(w_percent_as_local[data['Home'][i]][1])))
            w_percent_as_local[data['Home'][i]][0]= w_percent_as_local[data['Home'][i]][0]+1
            w_percent_as_local[data['Home'][i]][1]= w_percent_as_local[data['Home'][i]][1]+1
            w_percent_as_visitor[data['Visitor'][i]][1]=w_percent_as_visitor[data['Visitor'][i]][1]+1

    return w_percent_as_local,w_percent_as_visitor,w_percent_local,w_percent_visitor

win_percentage_as_local,win_percentage_as_visitor,w_percent_local,w_percent_visitor=w_perc_local_visitor(data,len(data))   


w_percent_as=[]
for i in range(len(w_percent_local)):
    
    w_percent_as=np.append(w_percent_as,"{0:.3f}".format(float(w_percent_local[i])-float(w_percent_visitor[i])))

print (win_percentage_as_local)
print(' ')
print (win_percentage_as_visitor)
#print (w_percent_v)
#print (w_percent_l)

data['%W as Home/Visitor']=w_percent_as
data

{'Toronto Raptors': [23, 32], 'Los Angeles Clippers': [25, 32], 'Charlotte Hornets': [10, 31], 'Indiana Pacers': [21, 32], 'Orlando Magic': [16, 31], 'Brooklyn Nets': [18, 32], 'Miami Heat': [27, 32], 'Philadelphia 76ers': [29, 31], 'Dallas Mavericks': [19, 34], 'San Antonio Spurs': [16, 30], 'Utah Jazz': [21, 31], 'Phoenix Suns': [13, 35], 'Portland Trail Blazers': [18, 32], 'Detroit Pistons': [11, 33], 'Houston Rockets': [22, 32], 'Golden State Warriors': [8, 34], 'Boston Celtics': [23, 32], 'Memphis Grizzlies': [18, 33], 'New Orleans Pelicans': [13, 32], 'Oklahoma City Thunder': [20, 33], 'Denver Nuggets': [25, 33], 'Sacramento Kings': [14, 31], 'Los Angeles Lakers': [23, 31], 'Milwaukee Bucks': [28, 31], 'Atlanta Hawks': [14, 34], 'New York Knicks': [11, 33], 'Chicago Bulls': [14, 34], 'Cleveland Cavaliers': [11, 36], 'Minnesota Timberwolves': [8, 32], 'Washington Wizards': [16, 32]}
 
{'New Orleans Pelicans': [15, 32], 'Los Angeles Lakers': [26, 32], 'Chicago Bulls': [8, 31], 'Det

Visitor  PTSV                  Home  PTSH  Overtime Attend.  \
0    New Orleans Pelicans   122       Toronto Raptors   130         1  20,787   
1      Los Angeles Lakers   102  Los Angeles Clippers   112         0  19,068   
2           Chicago Bulls   125     Charlotte Hornets   126         0  15,424   
3         Detroit Pistons   119        Indiana Pacers   110         0  17,923   
4     Cleveland Cavaliers    85         Orlando Magic    94         0  18,846   
..                    ...   ...                   ...   ...       ...     ...   
966         Brooklyn Nets   104    Los Angeles Lakers   102         0  18,997   
967       Detroit Pistons   106    Philadelphia 76ers   124         0  20,172   
968       New York Knicks   136         Atlanta Hawks   131         1  15,393   
969     Charlotte Hornets   109            Miami Heat    98         0    19,6   
970        Denver Nuggets    97      Dallas Mavericks   113         0  20,302   

     New Date      %W %W as Home/Visitor  
0    22/10/19   0.073              0.176  
1    22/10/19   0.049              0.180  
2    23/10/19  -0.079             -0.013  
3    23/10/19   0.134              0.239  
4    23/10/19   0.201              0.280  
..        ...     ...                ...  
966  10/03/20   0.330              0.412  
967  11/03/20   0.286              0.652  
968  11/03/20  -0.005              0.143  
969  11/03/20   0.297              0.507  
970  11/03/20  -0.081             -0.036  

[971 rows x 9 columns]

##### This function calculates back to back games of the teams, back to back games occurs when a team plays a game two consecutives days

In [13]:
def back_to_back(data,dim):
    last_game={}
    last_visitor=[]
    last_local=[]
    for i in range(dim):

        #initialize dictionary and arrays
        if data['Visitor'][i] not in last_game:
            last_game[data['Visitor'][i]]=data['New Date'][i]
            last_visitor=np.append(last_visitor,0)
        if data['Home'][i] not in last_game:
            last_game[data['Home'][i]]=data['New Date'][i]
            last_local=np.append(last_local,0)

        #normal actualitzation of the dictionary and arrays
        else:
            next_match=data['New Date'][i].split('/')

            last_match_visitor=last_game[data['Visitor'][i]].split('/')
            last_match_local=last_game[data['Home'][i]].split('/')
            if int(next_match[1])==int(last_match_visitor[1]):
                if int(next_match[0])-int(last_match_visitor[0])>=4:
                    last_visitor=np.append(last_visitor,0)
                else:
                    last_visitor=np.append(last_visitor,int(next_match[0])-int(last_match_visitor[0]))

            if int(next_match[1])==int(last_match_local[1]):
                if int(next_match[0])-int(last_match_local[0])>=4:
                    last_local=np.append(last_local,0)
                else:
                    last_local=np.append(last_local,int(next_match[0])-int(last_match_local[0]))

            if int(next_match[1])!=int(last_match_visitor[1]):
                if int(last_match_visitor[1]) in [10,12,1,3,5,7,8]:
                    if int(next_match[0])+31-int(last_match_visitor[0])>=4:
                        last_visitor=np.append(last_visitor,0)
                    else:
                        last_visitor=np.append(last_visitor,int(next_match[0])+31-int(last_match_visitor[0]))
                elif int(last_match_visitor[1]) in [11,4,6,9]:
                    if int(next_match[0])+30-int(last_match_visitor[0])>=4:
                        last_visitor=np.append(last_visitor,0)
                    else:
                        last_visitor=np.append(last_visitor,int(next_match[0])+30-int(last_match_visitor[0]))
                else:
                    if int(next_match[0])+29-int(last_match_visitor[0])>=4:
                        last_visitor=np.append(last_visitor,0)
                    else:
                        last_visitor=np.append(last_visitor,int(next_match[0])+29-int(last_match_visitor[0]))

            if int(next_match[1])!=int(last_match_local[1]):
                if int(last_match_local[1]) in [10,12,1,3,5,7,8]:
                    if int(next_match[0])+31-int(last_match_local[0])>=4:
                        last_local=np.append(last_local,0)
                    else:
                        last_local=np.append(last_local,int(next_match[0])+31-int(last_match_local[0]))
                elif int(last_match_local[1]) in [11,4,6,9]:
                    if int(next_match[0])+30-int(last_match_local[0])>=2:
                        last_local=np.append(last_local,0)
                    else:
                        last_local=np.append(last_local,int(next_match[0])+30-int(last_match_local[0]))
                else:
                    if int(next_match[0])+29-int(last_match_local[0])>=4:
                        last_local=np.append(last_local,0)
                    else:
                        last_local=np.append(last_local,int(next_match[0])+29-int(last_match_local[0]))
                    
            last_game[data['Visitor'][i]]=data['New Date'][i]
            last_game[data['Home'][i]]=data['New Date'][i]


    return last_game,last_local,last_visitor

dic_last_game,b2b_local,b2b_visitor=back_to_back(data,len(data))

b2b=[]
for i in range(len(b2b_local)):
    b2b=np.append(b2b,"{0:.2f}".format(float(b2b_local[i])-float(b2b_visitor[i])))

data['Back to back']=b2b
data

Visitor  PTSV                  Home  PTSH  Overtime Attend.  \
0    New Orleans Pelicans   122       Toronto Raptors   130         1  20,787   
1      Los Angeles Lakers   102  Los Angeles Clippers   112         0  19,068   
2           Chicago Bulls   125     Charlotte Hornets   126         0  15,424   
3         Detroit Pistons   119        Indiana Pacers   110         0  17,923   
4     Cleveland Cavaliers    85         Orlando Magic    94         0  18,846   
..                    ...   ...                   ...   ...       ...     ...   
966         Brooklyn Nets   104    Los Angeles Lakers   102         0  18,997   
967       Detroit Pistons   106    Philadelphia 76ers   124         0  20,172   
968       New York Knicks   136         Atlanta Hawks   131         1  15,393   
969     Charlotte Hornets   109            Miami Heat    98         0    19,6   
970        Denver Nuggets    97      Dallas Mavericks   113         0  20,302   

     New Date      %W %W as Home/Visitor Back to back  
0    22/10/19   0.073              0.176         0.00  
1    22/10/19   0.049              0.180         0.00  
2    23/10/19  -0.079             -0.013         0.00  
3    23/10/19   0.134              0.239         0.00  
4    23/10/19   0.201              0.280         0.00  
..        ...     ...                ...          ...  
966  10/03/20   0.330              0.412         0.00  
967  11/03/20   0.286              0.652        -3.00  
968  11/03/20  -0.005              0.143         1.00  
969  11/03/20   0.297              0.507         1.00  
970  11/03/20  -0.081             -0.036        -1.00  

[971 rows x 10 columns]

#### We want to know the division of NBA that belongs of every team is playing

In [14]:
Divisions={'Atlantic Division':['Toronto Raptors', 'Boston Celtics', 'Philadelphia 76ers', 'Brooklyn Nets', 'New York Knicks'],'Central Division':['Milwaukee Bucks', 'Indiana Pacers', 'Chicago Bulls','Detroit Pistons', 'Cleveland Cavaliers'],'SouthEast Division':['Miami Heat', 'Orlando Magic','Washington Wizards','Charlotte Hornets','Atlanta Hawks'],'NorWest Division':['Denver Nuggets','Utah Jazz','Oklahoma City Thunder','Portland Trail Blazers', 'Minnesota Timberwolves'],'Pacific Division':['Los Angeles Lakers','Los Angeles Clippers','Sacramento Kings','Phoenix Suns','Golden State Warriors'],'SouthWest Division':['Houston Rockets','Dallas Mavericks','Memphis Grizzlies','New Orleans Pelicans','San Antonio Spurs']}

def conf_div(data,dim):
    cod={'Atlantic Division':1,'Central Division':2,'SouthEast Division':3,'NorWest Division':4,'Pacific Division':5,'SouthWest Division':6}
    div_visitor=[]
    div_local=[]
    for i in range (dim):
        for key in Divisions:
            if data['Visitor'][i] in Divisions[key]:
                div_visitor=np.append(div_visitor,cod[key])
            if data['Home'][i] in Divisions[key]:
                div_local=np.append(div_local,cod[key])

    return div_visitor,div_local


div_v,div_l=conf_div(data,len(data))
#print((div_v))
#print((div_l))

data['Divisions Visitor']=div_v
data['Divisions Home']=div_l
data

Visitor  PTSV                  Home  PTSH  Overtime Attend.  \
0    New Orleans Pelicans   122       Toronto Raptors   130         1  20,787   
1      Los Angeles Lakers   102  Los Angeles Clippers   112         0  19,068   
2           Chicago Bulls   125     Charlotte Hornets   126         0  15,424   
3         Detroit Pistons   119        Indiana Pacers   110         0  17,923   
4     Cleveland Cavaliers    85         Orlando Magic    94         0  18,846   
..                    ...   ...                   ...   ...       ...     ...   
966         Brooklyn Nets   104    Los Angeles Lakers   102         0  18,997   
967       Detroit Pistons   106    Philadelphia 76ers   124         0  20,172   
968       New York Knicks   136         Atlanta Hawks   131         1  15,393   
969     Charlotte Hornets   109            Miami Heat    98         0    19,6   
970        Denver Nuggets    97      Dallas Mavericks   113         0  20,302   

     New Date      %W %W as Home/Visitor Back to back  Divisions Visitor  \
0    22/10/19   0.073              0.176         0.00                6.0   
1    22/10/19   0.049              0.180         0.00                5.0   
2    23/10/19  -0.079             -0.013         0.00                2.0   
3    23/10/19   0.134              0.239         0.00                2.0   
4    23/10/19   0.201              0.280         0.00                2.0   
..        ...     ...                ...          ...                ...   
966  10/03/20   0.330              0.412         0.00                1.0   
967  11/03/20   0.286              0.652        -3.00                2.0   
968  11/03/20  -0.005              0.143         1.00                1.0   
969  11/03/20   0.297              0.507         1.00                3.0   
970  11/03/20  -0.081             -0.036        -1.00                4.0   

     Divisions Home  
0               1.0  
1               5.0  
2               3.0  
3               2.0  
4               3.0  
..              ...  
966             5.0  
967             1.0  
968             3.0  
969             3.0  
970             6.0  

[971 rows x 12 columns]

#### We want to know the Conference that belongs every team is playing East=0, West=1

In [15]:
Conferences={'Eastern Conference':['Toronto Raptors', 'Boston Celtics', 'Philadelphia 76ers', 'Brooklyn Nets', 'New York Knicks','Milwaukee Bucks', 'Indiana Pacers', 'Chicago Bulls','Detroit Pistons', 'Cleveland Cavaliers','Miami Heat', 'Orlando Magic','Washington Wizards','Charlotte Hornets','Atlanta Hawks'],'Western Conference':['Denver Nuggets','Utah Jazz','Oklahoma City Thunder','Portland Trail Blazers', 'Minnesota Timberwolves','Los Angeles Lakers','Los Angeles Clippers','Sacramento Kings','Phoenix Suns','Golden State Warriors','Houston Rockets','Dallas Mavericks','Memphis Grizzlies','New Orleans Pelicans','San Antonio Spurs']}

def conf(data,dim):
    codi={'Western Conference':1,'Eastern Conference':0}
    conf_visitor=[]
    conf_local=[]
    for i in range (dim):
        for key in Conferences:
            if data['Visitor'][i] in Conferences[key]:
                conf_visitor=np.append(conf_visitor,codi[key])
            if data['Home'][i] in Conferences[key]:
                conf_local=np.append(conf_local,codi[key])

    return conf_visitor,conf_local


conf_v,conf_l=conf(data,len(data))
#print((conf_v))
#print((conf_l))
conf=[]
for i in range(len(conf_v)):
    conf=np.append(conf,"{0:.2f}".format(float(conf_l[i])-float(conf_v[i])))

data['Conference Visitor']=conf_v
data['Conference Home']=conf_l
data['Conference']=conf
data

Visitor  PTSV                  Home  PTSH  Overtime Attend.  \
0    New Orleans Pelicans   122       Toronto Raptors   130         1  20,787   
1      Los Angeles Lakers   102  Los Angeles Clippers   112         0  19,068   
2           Chicago Bulls   125     Charlotte Hornets   126         0  15,424   
3         Detroit Pistons   119        Indiana Pacers   110         0  17,923   
4     Cleveland Cavaliers    85         Orlando Magic    94         0  18,846   
..                    ...   ...                   ...   ...       ...     ...   
966         Brooklyn Nets   104    Los Angeles Lakers   102         0  18,997   
967       Detroit Pistons   106    Philadelphia 76ers   124         0  20,172   
968       New York Knicks   136         Atlanta Hawks   131         1  15,393   
969     Charlotte Hornets   109            Miami Heat    98         0    19,6   
970        Denver Nuggets    97      Dallas Mavericks   113         0  20,302   

     New Date      %W %W as Home/Visitor Back to back  Divisions Visitor  \
0    22/10/19   0.073              0.176         0.00                6.0   
1    22/10/19   0.049              0.180         0.00                5.0   
2    23/10/19  -0.079             -0.013         0.00                2.0   
3    23/10/19   0.134              0.239         0.00                2.0   
4    23/10/19   0.201              0.280         0.00                2.0   
..        ...     ...                ...          ...                ...   
966  10/03/20   0.330              0.412         0.00                1.0   
967  11/03/20   0.286              0.652        -3.00                2.0   
968  11/03/20  -0.005              0.143         1.00                1.0   
969  11/03/20   0.297              0.507         1.00                3.0   
970  11/03/20  -0.081             -0.036        -1.00                4.0   

     Divisions Home  Conference Visitor  Conference Home Conference  
0               1.0                 1.0              0.0      -1.00  
1               5.0                 1.0              1.0       0.00  
2               3.0                 0.0              0.0       0.00  
3               2.0                 0.0              0.0       0.00  
4               3.0                 0.0              0.0       0.00  
..              ...                 ...              ...        ...  
966             5.0                 0.0              1.0       1.00  
967             1.0                 0.0              0.0       0.00  
968             3.0                 0.0              0.0       0.00  
969             3.0                 0.0              0.0       0.00  
970             6.0                 1.0              1.0       0.00  

[971 rows x 15 columns]

#### Calculates the last N games number of wins of each team is similar to the last function but you can choose in a easier way how many games you want to consider   

In [16]:
def last_N_game(data,dim,N):
    last_N_games={}
    last_N_visitor=[]
    last_N_local=[]
    for i in range (dim):
        if data['Visitor'][i] not in last_N_games:
            last_N_games[data['Visitor'][i]]=[]

        if data['Home'][i] not in last_N_games:
            last_N_games[data['Home'][i]]=[]

        if int(data['PTSV'][i])>int(data['PTSH'][i]):

            if i<90:

                last_N_visitor=np.append(last_N_visitor, "{0:.2f}".format(float(Vegas_pred[data['Visitor'][i]])/float(82)))
                last_N_local=np.append(last_N_local, "{0:.2f}".format(float(Vegas_pred[data['Home'][i]])/float(82)))
            if i>=90:
                last_N_visitor=np.append(last_N_visitor, "{0:.2f}".format(last_N_games[data['Visitor'][i]].count('W')/len(last_N_games[data['Visitor'][i]])))
                last_N_local=np.append(last_N_local, "{0:.2f}".format(last_N_games[data['Home'][i]].count('W')/len(last_N_games[data['Home'][i]])))
            last_N_games[data['Visitor'][i]]= last_N_games[data['Visitor'][i]]+['W']
            last_N_games[data['Home'][i]]=last_N_games[data['Home'][i]]+['L']
            if len(last_N_games[data['Visitor'][i]])>=N:
                last_N_games[data['Visitor'][i]].pop(0)
                last_N_games[data['Visitor'][i]]=last_N_games[data['Visitor'][i]]
            if len(last_N_games[data['Home'][i]])>=N:
                last_N_games[data['Home'][i]].pop(0)
                last_N_games[data['Home'][i]]=last_N_games[data['Home'][i]]

        else:
            if i<90:

                last_N_visitor=np.append(last_N_visitor, "{0:.2f}".format(float(Vegas_pred[data['Visitor'][i]])/float(82)))
                last_N_local=np.append(last_N_local, "{0:.2f}".format(float(Vegas_pred[data['Home'][i]])/float(82)))
            if i>=90:
                last_N_visitor=np.append(last_N_visitor, "{0:.2f}".format(last_N_games[data['Visitor'][i]].count('W')/len(last_N_games[data['Visitor'][i]])))
                last_N_local=np.append(last_N_local,"{0:.2f}".format( last_N_games[data['Home'][i]].count('W')/len(last_N_games[data['Home'][i]])))


            last_N_games[data['Visitor'][i]]= last_N_games[data['Visitor'][i]]+['L']
            last_N_games[data['Home'][i]]=last_N_games[data['Home'][i]]+['W']

            if len(last_N_games[data['Visitor'][i]])>=N:
                last_N_games[data['Visitor'][i]].pop(0)
                last_N_games[data['Visitor'][i]]=last_N_games[data['Visitor'][i]]
            if len(last_N_games[data['Home'][i]])>=N:
                last_N_games[data['Home'][i]].pop(0)
                last_N_games[data['Home'][i]]=last_N_games[data['Home'][i]]

    
    return last_N_games,last_N_visitor,last_N_local

last_N_g,last_N_v,last_N_l=last_N_game(data,len(data),10)

last_N=[]
for i in range(len(last_N_v)):
    last_N=np.append(last_N,"{0:.2f}".format(float(last_N_l[i])-float(last_N_v[i])))

data['Last N']=last_N
data

Visitor  PTSV                  Home  PTSH  Overtime Attend.  \
0    New Orleans Pelicans   122       Toronto Raptors   130         1  20,787   
1      Los Angeles Lakers   102  Los Angeles Clippers   112         0  19,068   
2           Chicago Bulls   125     Charlotte Hornets   126         0  15,424   
3         Detroit Pistons   119        Indiana Pacers   110         0  17,923   
4     Cleveland Cavaliers    85         Orlando Magic    94         0  18,846   
..                    ...   ...                   ...   ...       ...     ...   
966         Brooklyn Nets   104    Los Angeles Lakers   102         0  18,997   
967       Detroit Pistons   106    Philadelphia 76ers   124         0  20,172   
968       New York Knicks   136         Atlanta Hawks   131         1  15,393   
969     Charlotte Hornets   109            Miami Heat    98         0    19,6   
970        Denver Nuggets    97      Dallas Mavericks   113         0  20,302   

     New Date      %W %W as Home/Visitor Back to back  Divisions Visitor  \
0    22/10/19   0.073              0.176         0.00                6.0   
1    22/10/19   0.049              0.180         0.00                5.0   
2    23/10/19  -0.079             -0.013         0.00                2.0   
3    23/10/19   0.134              0.239         0.00                2.0   
4    23/10/19   0.201              0.280         0.00                2.0   
..        ...     ...                ...          ...                ...   
966  10/03/20   0.330              0.412         0.00                1.0   
967  11/03/20   0.286              0.652        -3.00                2.0   
968  11/03/20  -0.005              0.143         1.00                1.0   
969  11/03/20   0.297              0.507         1.00                3.0   
970  11/03/20  -0.081             -0.036        -1.00                4.0   

     Divisions Home  Conference Visitor  Conference Home Conference Last N  
0               1.0                 1.0              0.0      -1.00   0.07  
1               5.0                 1.0              1.0       0.00   0.05  
2               3.0                 0.0              0.0       0.00  -0.08  
3               2.0                 0.0              0.0       0.00   0.13  
4               3.0                 0.0              0.0       0.00   0.20  
..              ...                 ...              ...        ...    ...  
966             5.0                 0.0              1.0       1.00   0.45  
967             1.0                 0.0              0.0       0.00   0.33  
968             3.0                 0.0              0.0       0.00   0.11  
969             3.0                 0.0              0.0       0.00   0.34  
970             6.0                 1.0              1.0       0.00   0.00  

[971 rows x 16 columns]

#### We calculate the net rating of each team in the list and how is evolving in every match, the net rating it's the difference between points for the team and against the team, divided with the number of games played

In [17]:
def net_rating(data,dim):
    net_rat={}
    net_rat_visitor=[]
    net_rat_local=[]
    for i in range(dim):
        if data['Visitor'][i] not in net_rat:
            net_rat[data['Visitor'][i]]=[0,1]
            net_rat_visitor=np.append(net_rat_visitor,"{0:.2f}".format(0.0))
        if data['Home'][i] not in net_rat:
            net_rat[data['Home'][i]]=[0,1]   
            net_rat_local=np.append(net_rat_local, "{0:.2f}".format(0.0))
        else:
            dif=int(data['PTSV'][i])-int(data['PTSH'][i])

            net_rat_visitor=np.append(net_rat_visitor,"{0:.2f}".format(int(net_rat[data['Visitor'][i]][0])/int(net_rat[data['Visitor'][i]][1])))
            net_rat_local=np.append(net_rat_local,"{0:.2f}".format(int(net_rat[data['Home'][i]][0])/int(net_rat[data['Home'][i]][1])))
            if dif<0:
                dif=dif*-1
            if int(data['PTSV'][i])>int(data['PTSH'][i]):
                net_rat[data['Visitor'][i]][0]=net_rat[data['Visitor'][i]][0]+dif
                net_rat[data['Visitor'][i]][1]=net_rat[data['Visitor'][i]][1]+1
                net_rat[data['Home'][i]][0]=net_rat[data['Home'][i]][0]+(dif*-1)
                net_rat[data['Home'][i]][1]=net_rat[data['Home'][i]][1]+1
            if int(data['PTSV'][i])<int(data['PTSH'][i]):
                net_rat[data['Visitor'][i]][0]=net_rat[data['Visitor'][i]][0]+(dif*-1)
                net_rat[data['Visitor'][i]][1]=net_rat[data['Visitor'][i]][1]+1
                net_rat[data['Home'][i]][0]=net_rat[data['Home'][i]][0]+dif
                net_rat[data['Home'][i]][1]=net_rat[data['Home'][i]][1]+1

    return net_rat,net_rat_visitor,net_rat_local

net_rating,net_rating_visitor,net_rating_local=net_rating(data,len(data))


net_rating=[]
for i in range(len(net_rating_local)):
    net_rating=np.append(net_rating,"{0:.3f}".format(float(net_rating_local[i])-float(net_rating_visitor[i])))

data['Net Rating']=net_rating
data

Visitor  PTSV                  Home  PTSH  Overtime Attend.  \
0    New Orleans Pelicans   122       Toronto Raptors   130         1  20,787   
1      Los Angeles Lakers   102  Los Angeles Clippers   112         0  19,068   
2           Chicago Bulls   125     Charlotte Hornets   126         0  15,424   
3         Detroit Pistons   119        Indiana Pacers   110         0  17,923   
4     Cleveland Cavaliers    85         Orlando Magic    94         0  18,846   
..                    ...   ...                   ...   ...       ...     ...   
966         Brooklyn Nets   104    Los Angeles Lakers   102         0  18,997   
967       Detroit Pistons   106    Philadelphia 76ers   124         0  20,172   
968       New York Knicks   136         Atlanta Hawks   131         1  15,393   
969     Charlotte Hornets   109            Miami Heat    98         0    19,6   
970        Denver Nuggets    97      Dallas Mavericks   113         0  20,302   

     New Date      %W %W as Home/Visitor Back to back  Divisions Visitor  \
0    22/10/19   0.073              0.176         0.00                6.0   
1    22/10/19   0.049              0.180         0.00                5.0   
2    23/10/19  -0.079             -0.013         0.00                2.0   
3    23/10/19   0.134              0.239         0.00                2.0   
4    23/10/19   0.201              0.280         0.00                2.0   
..        ...     ...                ...          ...                ...   
966  10/03/20   0.330              0.412         0.00                1.0   
967  11/03/20   0.286              0.652        -3.00                2.0   
968  11/03/20  -0.005              0.143         1.00                1.0   
969  11/03/20   0.297              0.507         1.00                3.0   
970  11/03/20  -0.081             -0.036        -1.00                4.0   

     Divisions Home  Conference Visitor  Conference Home Conference Last N  \
0               1.0                 1.0              0.0      -1.00   0.07   
1               5.0                 1.0              1.0       0.00   0.05   
2               3.0                 0.0              0.0       0.00  -0.08   
3               2.0                 0.0              0.0       0.00   0.13   
4               3.0                 0.0              0.0       0.00   0.20   
..              ...                 ...              ...        ...    ...   
966             5.0                 0.0              1.0       1.00   0.45   
967             1.0                 0.0              0.0       0.00   0.33   
968             3.0                 0.0              0.0       0.00   0.11   
969             3.0                 0.0              0.0       0.00   0.34   
970             6.0                 1.0              1.0       0.00   0.00   

    Net Rating  
0        0.000  
1        0.000  
2        0.000  
3        0.000  
4        0.000  
..         ...  
966      8.400  
967      5.230  
968     -1.410  
969     10.210  
970      2.650  

[971 rows x 17 columns]

#### Sets 1 if the team played an OT last match if not it sets 0

In [18]:
def OT_last_match(data,dim):
    overtime={}
    overtime_local=[]
    overtime_visitor=[]
    for i in range (dim):
        if  data['Home'][i] not in overtime:
            overtime[data['Home'][i]]=0
        if  data['Visitor'][i] not in overtime:
            overtime[data['Visitor'][i]]=0
        overtime_local=np.append(overtime_local,overtime[data['Home'][i]])
        overtime_visitor=np.append(overtime_visitor,overtime[data['Visitor'][i]])
        if data['Overtime'][i]==1:
            if overtime[data['Home'][i]]==0:
                overtime[data['Home'][i]]=overtime[data['Home'][i]]+1
            if overtime[data['Visitor'][i]]==0:
                overtime[data['Visitor'][i]]=overtime[data['Visitor'][i]]+1

        if data['Overtime'][i]==0 and overtime[data['Home'][i]]==1:
            overtime[data['Home'][i]]=overtime[data['Home'][i]]-1
        if data['Overtime'][i]==0 and overtime[data['Visitor'][i]]==1:
            overtime[data['Visitor'][i]]=overtime[data['Visitor'][i]]-1

    return(overtime,overtime_local,overtime_visitor)

OTs,OT_local,OT_visit=OT_last_match(data,len(data))
#print(OTs,OT_local,OT_visit)

OT=[]
for i in range(len(OT_local)):
    OT=np.append(OT,"{0:.2f}".format(float(OT_local[i])-float(OT_visit[i])))

data['OT last match']=OT
data

Visitor  PTSV                  Home  PTSH  Overtime Attend.  \
0    New Orleans Pelicans   122       Toronto Raptors   130         1  20,787   
1      Los Angeles Lakers   102  Los Angeles Clippers   112         0  19,068   
2           Chicago Bulls   125     Charlotte Hornets   126         0  15,424   
3         Detroit Pistons   119        Indiana Pacers   110         0  17,923   
4     Cleveland Cavaliers    85         Orlando Magic    94         0  18,846   
..                    ...   ...                   ...   ...       ...     ...   
966         Brooklyn Nets   104    Los Angeles Lakers   102         0  18,997   
967       Detroit Pistons   106    Philadelphia 76ers   124         0  20,172   
968       New York Knicks   136         Atlanta Hawks   131         1  15,393   
969     Charlotte Hornets   109            Miami Heat    98         0    19,6   
970        Denver Nuggets    97      Dallas Mavericks   113         0  20,302   

     New Date      %W %W as Home/Visitor Back to back  Divisions Visitor  \
0    22/10/19   0.073              0.176         0.00                6.0   
1    22/10/19   0.049              0.180         0.00                5.0   
2    23/10/19  -0.079             -0.013         0.00                2.0   
3    23/10/19   0.134              0.239         0.00                2.0   
4    23/10/19   0.201              0.280         0.00                2.0   
..        ...     ...                ...          ...                ...   
966  10/03/20   0.330              0.412         0.00                1.0   
967  11/03/20   0.286              0.652        -3.00                2.0   
968  11/03/20  -0.005              0.143         1.00                1.0   
969  11/03/20   0.297              0.507         1.00                3.0   
970  11/03/20  -0.081             -0.036        -1.00                4.0   

     Divisions Home  Conference Visitor  Conference Home Conference Last N  \
0               1.0                 1.0              0.0      -1.00   0.07   
1               5.0                 1.0              1.0       0.00   0.05   
2               3.0                 0.0              0.0       0.00  -0.08   
3               2.0                 0.0              0.0       0.00   0.13   
4               3.0                 0.0              0.0       0.00   0.20   
..              ...                 ...              ...        ...    ...   
966             5.0                 0.0              1.0       1.00   0.45   
967             1.0                 0.0              0.0       0.00   0.33   
968             3.0                 0.0              0.0       0.00   0.11   
969             3.0                 0.0              0.0       0.00   0.34   
970             6.0                 1.0              1.0       0.00   0.00   

    Net Rating OT last match  
0        0.000          0.00  
1        0.000          0.00  
2        0.000          0.00  
3        0.000          0.00  
4        0.000          0.00  
..         ...           ...  
966      8.400          0.00  
967      5.230          0.00  
968     -1.410          1.00  
969     10.210         -1.00  
970      2.650          0.00  

[971 rows x 18 columns]

#### The following functions gives the data of win shares,PER (similar to "Valoració" in ACB) and value of replacement(similar to +/- stat) of every important player in the league. We have choosen the 100th best players in every stat and we have collected their stats. The function return the dictionaries that contain the information of each player

In [19]:
team_capitals={'Toronto Raptors':'TOR', 'Boston Celtics':'BOS', 'Philadelphia 76ers':'PHI', 'Brooklyn Nets':'BRK', 'New York Knicks':'NYK','Milwaukee Bucks':'MIL', 'Indiana Pacers':'IND', 'Chicago Bulls':'CHI','Detroit Pistons':'DET', 'Cleveland Cavaliers':'CLE','Miami Heat':'MIA', 'Orlando Magic':'ORL','Washington Wizards':'WAS','Charlotte Hornets':'CHO','Atlanta Hawks':'ATL','Denver Nuggets':'DEN','Utah Jazz':'UTA','Oklahoma City Thunder':'OKC','Portland Trail Blazers':'POR', 'Minnesota Timberwolves':'MIN','Los Angeles Lakers':'LAL','Los Angeles Clippers':'LAC','Sacramento Kings':'SAC','Phoenix Suns':'PHO','Golden State Warriors':'GSW','Houston Rockets':'HOU','Dallas Mavericks':'DAL','Memphis Grizzlies':'MEM','New Orleans Pelicans':'NOP','San Antonio Spurs':'SAS'}

def player_stat(data,dim):
    stars={}
    for i in range (dim):
        if data['STAR'][i]==1:
            stars[data['Player'][i]]=data['STAR'][i]
    return stars


stars=player_stat(data_players,len(data_players))
print(stars)

{'James Harden': 1, 'Giannis Antetokounmpo': 1, 'LeBron James': 1, 'Nikola Jokic': 1, 'Anthony Davis': 1, 'Damian Lillard': 1, 'Luka Doncic': 1, 'Kawhi Leonard': 1, 'Jimmy Butler': 1, 'Chris Paul': 1, 'Rudy Gobert': 1, 'Jayson Tatum': 1, 'Ben Simmons': 1, 'Karl-Anthony Towns': 1, 'Khris Middleton': 1, 'Nikola Vucevic': 1, 'Kemba Walker': 1, 'Bradley Beal': 1, 'Kyle Lowry': 1, 'Joel Embiid': 1, 'Danilo Gallinari': 1, 'Devin Booker': 1, 'Jrue Holiday': 1, 'Donovan Mitchell': 1, 'Steven Adams': 1, 'Paul George': 1, 'Al Horford': 1, 'Pascal Siakam': 1, 'DeMar DeRozan': 1, 'Kevin Love': 1, 'Russell Westbrook': 1, 'Eric Bledsoe': 1, 'Kyrie Irving': 1, 'CJ McCollum': 1, 'LaMarcus Aldridge': 1, 'Tobias Harris': 1, 'Andre Drummond': 1, "De'Aaron Fox": 1, 'Jamal Murray': 1, 'Kristaps Porzingis': 1, "D'Angelo Russell": 1, 'Marc Gasol': 1, 'Myles Turner': 1, 'Paul Millsap': 1, 'Mike Conley': 1, 'Draymond Green': 1, 'Stephen Curry': 1, 'Victor Oladipo': 1, 'Gary Harris': 1, 'Blake Griffin': 1}


#### Gives a dict with all the players in VORP dic, PER dic, with the matches that they have played during the season

In [20]:
def dic_games_played(dic):
    months={'OCT':'10','NOV':'11','DEC':'12','JAN':'01','FEB':'02','MAR':'03','APR':'04','MAY':'05','JUN':'06','JUL':'07','AUG':'08','SEP':'09'}
    games_players={}
    for key in dic:
        games=[]
        players= pd.read_csv("NBA_project/players_19_20/games_"+key+".csv")

        for i in range (len(players)):
            match=players['GAME_DATE'][i]
            match=match.split(',')
            m=match[0].split(' ')
            m=str(m[1])+'/'+str(months[m[0]])+'/'+match[1][3::]
            games=games+[m]
        games_players[key]=games
    return games_players

dic_games_played=dic_games_played(stars)
print(dic_games_played)

{'James Harden': ['10/03/20', '08/03/20', '07/03/20', '05/03/20', '02/03/20', '29/02/20', '26/02/20', '24/02/20', '22/02/20', '20/02/20', '11/02/20', '09/02/20', '07/02/20', '06/02/20', '04/02/20', '02/02/20', '31/01/20', '29/01/20', '24/01/20', '22/01/20', '20/01/20', '18/01/20', '15/01/20', '14/01/20', '11/01/20', '09/01/20', '08/01/20', '03/01/20', '31/12/19', '28/12/19', '25/12/19', '23/12/19', '21/12/19', '19/12/19', '16/12/19', '14/12/19', '13/12/19', '11/12/19', '09/12/19', '07/12/19', '05/12/19', '03/12/19', '30/11/19', '27/11/19', '24/11/19', '22/11/19', '20/11/19', '18/11/19', '16/11/19', '15/11/19', '13/11/19', '11/11/19', '09/11/19', '06/11/19', '04/11/19', '03/11/19', '01/11/19', '30/10/19', '28/10/19', '26/10/19', '24/10/19'], 'Giannis Antetokounmpo': ['06/03/20', '04/03/20', '02/03/20', '01/03/20', '28/02/20', '25/02/20', '24/02/20', '22/02/20', '20/02/20', '08/02/20', '06/02/20', '04/02/20', '02/02/20', '31/01/20', '24/01/20', '20/01/20', '18/01/20', '16/01/20', '14/01/

#### This function makes a dictionary that contains the player as the key and the team that belongs this player as the value.

In [21]:
def player_team(data,dim,dic):
    players_teams={}
    for i in range(dim):
        last_team=[]
        if data['Player'][i] in stars and data['Tm'][i] == 'TOT':

            j=0
            players= pd.read_csv("NBA_project/players_19_20/games_"+data['Player'][i]+".csv")
            while j+1 < (len(players)):
                j=j+1
                if players['MATCHUP'][j][:3]!=players['MATCHUP'][j-1][:3]:
                    last_team=last_team+[players['MATCHUP'][j-1][:3],players['MATCHUP'][j][:3],players['GAME_DATE'][j]]

            players_teams[data['Player'][i]]=last_team
            
        
        if data['Player'][i] in stars and data['Tm'][i] not in players_teams and data['Tm'][i] != 'TOT' :
            players_teams[data['Player'][i]]=[data['Tm'][i]]

    return players_teams


play_team=player_team(data_players,len(data_players),dic_games_played)
print(play_team)



{'James Harden': ['HOU'], 'Giannis Antetokounmpo': ['MIL'], 'LeBron James': ['LAL'], 'Nikola Jokic': ['DEN'], 'Anthony Davis': ['LAL'], 'Damian Lillard': ['POR'], 'Luka Doncic': ['DAL'], 'Kawhi Leonard': ['LAC'], 'Jimmy Butler': ['MIA'], 'Chris Paul': ['OKC'], 'Rudy Gobert': ['UTA'], 'Jayson Tatum': ['BOS'], 'Ben Simmons': ['PHI'], 'Karl-Anthony Towns': ['MIN'], 'Khris Middleton': ['MIL'], 'Nikola Vucevic': ['ORL'], 'Kemba Walker': ['BOS'], 'Bradley Beal': ['WAS'], 'Kyle Lowry': ['TOR'], 'Joel Embiid': ['PHI'], 'Danilo Gallinari': ['OKC'], 'Devin Booker': ['PHO'], 'Jrue Holiday': ['NOP'], 'Donovan Mitchell': ['UTA'], 'Steven Adams': ['OKC'], 'Paul George': ['LAC'], 'Al Horford': ['PHI'], 'Pascal Siakam': ['TOR'], 'DeMar DeRozan': ['SAS'], 'Kevin Love': ['CLE'], 'Russell Westbrook': ['HOU'], 'Eric Bledsoe': ['MIL'], 'Kyrie Irving': ['BRK'], 'CJ McCollum': ['POR'], 'LaMarcus Aldridge': ['SAS'], 'Tobias Harris': ['PHI'], 'Andre Drummond': ['CLE', 'DET', 'FEB 05, 2020'], "De'Aaron Fox": ['

##### Comparation of dates

In [22]:

def date_before(date1,date2):
    date1=date1.split('/')
    months={'OCT':'10','NOV':'11','DEC':'12','JAN':'01','FEB':'02','MAR':'03','APR':'04','MAY':'05','JUN':'06','JUL':'07','SEP':'09'} 
    # importing datetime module 
    import datetime
    match=date2
    match=match.split(',')
    m=match[0].split(' ')
    m=[int(m[1]),int(months[m[0]])]
    # date in yyyy/mm/dd format 
    if date1[1] in ['10','11','12']:
        year1=2019
    if str(m[1]) in ['10','11','12']:
        year2=2019
    if date1[1] in ['01','02','03','04']:
        year1=2020
    if str(m[1]) in ['1','2','3','4']:
        year2=2020
    d1 = datetime.datetime(year1, int(date1[1]), int(date1[0])) 
    d2 = datetime.datetime(year2, m[1],m[0] ) 
    return d1<d2

#### This function sets the value of replacement,true shooting,PER,win shares, BPM, %usage,true rebounding, true shooting, for each team in every match of the players that are not playing the match

In [23]:
def stars_rest_inj(data_players,dim_players,players,data,dim,stat):
    stars_inj_local=[]
    stars_inj_visitor=[]
    
    for i in range (dim):
      
        value_local=0
        value_visitor=0
        for key in players:
            
            if len(play_team[key])==3:
                
                if date_before(data['New Date'][i],play_team[key][2]) and data['New Date'][i] not in players[key]and play_team[key][1] == team_capitals[data['Home'][i]]:
                   
                    value_local=value_local+stat[key]
                if date_before(data['New Date'][i],play_team[key][2])==False and data['New Date'][i] not in players[key] and play_team[key][0] == team_capitals[data['Home'][i]]:
                   
                    value_local=value_local+stat[key]

                if date_before(data['New Date'][i],play_team[key][2]) and data['New Date'][i] not in players[key]and play_team[key][1] == team_capitals[data['Visitor'][i]]:
                   
                    value_visitor=value_local+stat[key]
                if date_before(data['New Date'][i],play_team[key][2])==False and data['New Date'][i] not in players[key] and play_team[key][0] == team_capitals[data['Visitor'][i]]:
                 
                    value_visitor=value_local+stat[key]
            else:
                
                if data['New Date'][i] not in players[key] and play_team[key][0] == team_capitals[data['Home'][i]]:
                                   
                    value_local=value_local+stat[key]
                
                if data['New Date'][i] not in players[key] and play_team[key][0] == team_capitals[data['Visitor'][i]]:
                    value_visitor=value_visitor+stat[key]

        stars_inj_local=np.append(stars_inj_local,"{0:.2f}".format(value_local))
        stars_inj_visitor=np.append(stars_inj_visitor,"{0:.2f}".format(value_visitor))


    return stars_inj_local,stars_inj_visitor

stars_inj_loc,stars_inj_visit=stars_rest_inj(data_players,len(data_players),dic_games_played,data,len(data),stars)
stars_inj=[]
for j in range(len(stars_inj_loc)):
    stars_inj=np.append(stars_inj,"{0:.2f}".format(float(stars_inj_loc[j])-float(stars_inj_visit[j])))
data['Rest/Injuries Visitor Stars']=stars_inj_visit
data['Rest/Injuries Home Stars']=stars_inj_loc
data['Rest/Injuries Stars']=stars_inj
data
#print(stars_inj_loc)
#print(stars_inj_visit)

Visitor  PTSV                  Home  PTSH  Overtime Attend.  \
0    New Orleans Pelicans   122       Toronto Raptors   130         1  20,787   
1      Los Angeles Lakers   102  Los Angeles Clippers   112         0  19,068   
2           Chicago Bulls   125     Charlotte Hornets   126         0  15,424   
3         Detroit Pistons   119        Indiana Pacers   110         0  17,923   
4     Cleveland Cavaliers    85         Orlando Magic    94         0  18,846   
..                    ...   ...                   ...   ...       ...     ...   
966         Brooklyn Nets   104    Los Angeles Lakers   102         0  18,997   
967       Detroit Pistons   106    Philadelphia 76ers   124         0  20,172   
968       New York Knicks   136         Atlanta Hawks   131         1  15,393   
969     Charlotte Hornets   109            Miami Heat    98         0    19,6   
970        Denver Nuggets    97      Dallas Mavericks   113         0  20,302   

     New Date      %W %W as Home/Visitor Back to back  ...  Divisions Home  \
0    22/10/19   0.073              0.176         0.00  ...             1.0   
1    22/10/19   0.049              0.180         0.00  ...             5.0   
2    23/10/19  -0.079             -0.013         0.00  ...             3.0   
3    23/10/19   0.134              0.239         0.00  ...             2.0   
4    23/10/19   0.201              0.280         0.00  ...             3.0   
..        ...     ...                ...          ...  ...             ...   
966  10/03/20   0.330              0.412         0.00  ...             5.0   
967  11/03/20   0.286              0.652        -3.00  ...             1.0   
968  11/03/20  -0.005              0.143         1.00  ...             3.0   
969  11/03/20   0.297              0.507         1.00  ...             3.0   
970  11/03/20  -0.081             -0.036        -1.00  ...             6.0   

     Conference Visitor  Conference Home  Conference Last N Net Rating  \
0                   1.0              0.0       -1.00   0.07      0.000   
1                   1.0              1.0        0.00   0.05      0.000   
2                   0.0              0.0        0.00  -0.08      0.000   
3                   0.0              0.0        0.00   0.13      0.000   
4                   0.0              0.0        0.00   0.20      0.000   
..                  ...              ...         ...    ...        ...   
966                 0.0              1.0        1.00   0.45      8.400   
967                 0.0              0.0        0.00   0.33      5.230   
968                 0.0              0.0        0.00   0.11     -1.410   
969                 0.0              0.0        0.00   0.34     10.210   
970                 1.0              1.0        0.00   0.00      2.650   

    OT last match Rest/Injuries Visitor Stars Rest/Injuries Home Stars  \
0            0.00                        0.00                     0.00   
1            0.00                        0.00                     1.00   
2            0.00                        0.00                     0.00   
3            0.00                        1.00                     1.00   
4            0.00                        0.00                     0.00   
..            ...                         ...                      ...   
966          0.00                        1.00                     0.00   
967          0.00                        1.00                     1.00   
968          1.00                        0.00                     0.00   
969         -1.00                        0.00                     1.00   
970          0.00                        0.00                     1.00   

    Rest/Injuries Stars  
0                  0.00  
1                  1.00  
2                  0.00  
3                  0.00  
4                  0.00  
..                  ...  
966               -1.00  
967                0.00  
968                0.00  
969                1.00  
970                1.00  

[971 rows x 21 co

#### Gives the win streak or loss streak if its bigger than n, win streak in positive, loss streak in negative

In [24]:
def streak(data,dim,N,n):
    streak_N_games={}
    streak_N_visitor=[]
    streak_N_local=[]
    for i in range (dim):

        if data['Visitor'][i] not in streak_N_games:
           streak_N_games[data['Visitor'][i]]=[0]
           streak_N_visitor=np.append(streak_N_visitor, 0)

        if data['Home'][i] not in streak_N_games:
            streak_N_games[data['Home'][i]]=[0]
            streak_N_local=np.append(streak_N_local, 0)
        wins=0
        lost=0
        if streak_N_games[data['Visitor'][i]][-1]=='W':

            for j in range(len(streak_N_games[data['Visitor'][i]])):
                if streak_N_games[data['Visitor'][i]][-j-1]=='W':
                    wins=wins+1
                else:
                    break
            if wins>=n:
                streak_N_visitor=np.append(streak_N_visitor, wins)
            else:
                streak_N_visitor=np.append(streak_N_visitor, 0)

        if streak_N_games[data['Visitor'][i]][-1]=='L':

            for j in range(len(streak_N_games[data['Visitor'][i]])):
                if streak_N_games[data['Visitor'][i]][-j-1]=='L':
                    lost=lost+1
                else:
                    break
            if lost>=n:
                streak_N_visitor=np.append(streak_N_visitor, -lost)
            else:
                streak_N_visitor=np.append(streak_N_visitor, 0)     
        wins=0
        lost=0
        if streak_N_games[data['Home'][i]][-1]=='W':

            for j in range(len(streak_N_games[data['Home'][i]])):
                if streak_N_games[data['Home'][i]][-j-1]=='W':
                    wins=wins+1
                else:
                    break
            if wins>=n:
                streak_N_local=np.append(streak_N_local, wins)
            else:
                streak_N_local=np.append(streak_N_local, 0)

        if streak_N_games[data['Home'][i]][-1]=='L':

            for j in range(len(streak_N_games[data['Home'][i]])):
                if streak_N_games[data['Home'][i]][-j-1]=='L':
                    lost=lost+1
                else:
                    break
            if lost>=n:
                streak_N_local=np.append(streak_N_local, -lost)
            else:
                streak_N_local=np.append(streak_N_local, 0)
        

        if int(data['PTSV'][i])>int(data['PTSH'][i]):
            streak_N_games[data['Visitor'][i]]= streak_N_games[data['Visitor'][i]]+['W']
            streak_N_games[data['Home'][i]]=streak_N_games[data['Home'][i]]+['L']
            if len(streak_N_games[data['Visitor'][i]])>=N:
                streak_N_games[data['Visitor'][i]].pop(0)
                streak_N_games[data['Visitor'][i]]=streak_N_games[data['Visitor'][i]]
            if len(streak_N_games[data['Home'][i]])>=N:
                streak_N_games[data['Home'][i]].pop(0)
                streak_N_games[data['Home'][i]]=streak_N_games[data['Home'][i]]

        else:
            streak_N_games[data['Visitor'][i]]= streak_N_games[data['Visitor'][i]]+['L']
            streak_N_games[data['Home'][i]]=streak_N_games[data['Home'][i]]+['W']

            if len(streak_N_games[data['Visitor'][i]])>=N:
                streak_N_games[data['Visitor'][i]].pop(0)
                streak_N_games[data['Visitor'][i]]=streak_N_games[data['Visitor'][i]]
            if len(streak_N_games[data['Home'][i]])>=N:
                streak_N_games[data['Home'][i]].pop(0)
                streak_N_games[data['Home'][i]]=streak_N_games[data['Home'][i]]

    
    return streak_N_games,streak_N_visitor,streak_N_local

streak_N_g,streak_N_v,streak_N_l=streak(data,len(data),30,8)

streak_N=[]
for i in range(len(streak_N_l)):
    streak_N=np.append(streak_N,"{0:.2f}".format(float(streak_N_l[i])-float(streak_N_v[i])))

data['Win Streak']=streak_N
data


Visitor  PTSV                  Home  PTSH  Overtime Attend.  \
0    New Orleans Pelicans   122       Toronto Raptors   130         1  20,787   
1      Los Angeles Lakers   102  Los Angeles Clippers   112         0  19,068   
2           Chicago Bulls   125     Charlotte Hornets   126         0  15,424   
3         Detroit Pistons   119        Indiana Pacers   110         0  17,923   
4     Cleveland Cavaliers    85         Orlando Magic    94         0  18,846   
..                    ...   ...                   ...   ...       ...     ...   
966         Brooklyn Nets   104    Los Angeles Lakers   102         0  18,997   
967       Detroit Pistons   106    Philadelphia 76ers   124         0  20,172   
968       New York Knicks   136         Atlanta Hawks   131         1  15,393   
969     Charlotte Hornets   109            Miami Heat    98         0    19,6   
970        Denver Nuggets    97      Dallas Mavericks   113         0  20,302   

     New Date      %W %W as Home/Visitor Back to back  ...  \
0    22/10/19   0.073              0.176         0.00  ...   
1    22/10/19   0.049              0.180         0.00  ...   
2    23/10/19  -0.079             -0.013         0.00  ...   
3    23/10/19   0.134              0.239         0.00  ...   
4    23/10/19   0.201              0.280         0.00  ...   
..        ...     ...                ...          ...  ...   
966  10/03/20   0.330              0.412         0.00  ...   
967  11/03/20   0.286              0.652        -3.00  ...   
968  11/03/20  -0.005              0.143         1.00  ...   
969  11/03/20   0.297              0.507         1.00  ...   
970  11/03/20  -0.081             -0.036        -1.00  ...   

     Conference Visitor  Conference Home  Conference  Last N Net Rating  \
0                   1.0              0.0       -1.00    0.07      0.000   
1                   1.0              1.0        0.00    0.05      0.000   
2                   0.0              0.0        0.00   -0.08      0.000   
3                   0.0              0.0        0.00    0.13      0.000   
4                   0.0              0.0        0.00    0.20      0.000   
..                  ...              ...         ...     ...        ...   
966                 0.0              1.0        1.00    0.45      8.400   
967                 0.0              0.0        0.00    0.33      5.230   
968                 0.0              0.0        0.00    0.11     -1.410   
969                 0.0              0.0        0.00    0.34     10.210   
970                 1.0              1.0        0.00    0.00      2.650   

    OT last match Rest/Injuries Visitor Stars Rest/Injuries Home Stars  \
0            0.00                        0.00                     0.00   
1            0.00                        0.00                     1.00   
2            0.00                        0.00                     0.00   
3            0.00                        1.00                     1.00   
4            0.00                        0.00                     0.00   
..            ...                         ...                      ...   
966          0.00                        1.00                     0.00   
967          0.00                        1.00                     1.00   
968          1.00                        0.00                     0.00   
969         -1.00                        0.00                     1.00   
970          0.00                        0.00                     1.00   

    Rest/Injuries Stars Win Streak  
0                  0.00       0.00  
1                  1.00       0.00  
2                  0.00       0.00  
3                  0.00       0.00  
4                  0.00       0.00  
..                  ...        ...  
966               -1.00       0.00  
967                0.00       0.00  
968                0.00       0.00  
969                1.00       0.00  
970                1.00       0.00  

[971 rows x 22 columns]

#### Give the bets on every match of the season for the local and the visitor team

In [25]:
def bets(data_bets,dim_bets,data):
    bet_visitor=[]
    bet_local=[]
    for i in range(len(data)):
       
        for j in range (dim_bets):
            
            if data['New Date'][i] in data_bets['NEW DATE'][j] and data['Home'][i] in data_bets['HOME'][j] and data['Visitor'][i] in data_bets['VISITOR'][j]:
                #print(i,s)
              
                bet_visitor=np.append(bet_visitor,"{0:.2f}".format(data_bets['BET VISITOR'][j]))
                bet_local=np.append(bet_local,"{0:.2f}".format(data_bets['BET HOME'][j]))
                break


    return bet_visitor,bet_local

bet_v,bet_l=bets(data_bets,len(data_bets),data)

data['Bet Visitor']=bet_v
data['Bet Home']=bet_l
bets=[]
for i in range(len(bet_v)):
    bets=np.append(bets,"{0:.2f}".format(float(bet_l[i])-float(bet_v[i])))

data['Bets']=bets
data


Visitor  PTSV                  Home  PTSH  Overtime Attend.  \
0    New Orleans Pelicans   122       Toronto Raptors   130         1  20,787   
1      Los Angeles Lakers   102  Los Angeles Clippers   112         0  19,068   
2           Chicago Bulls   125     Charlotte Hornets   126         0  15,424   
3         Detroit Pistons   119        Indiana Pacers   110         0  17,923   
4     Cleveland Cavaliers    85         Orlando Magic    94         0  18,846   
..                    ...   ...                   ...   ...       ...     ...   
966         Brooklyn Nets   104    Los Angeles Lakers   102         0  18,997   
967       Detroit Pistons   106    Philadelphia 76ers   124         0  20,172   
968       New York Knicks   136         Atlanta Hawks   131         1  15,393   
969     Charlotte Hornets   109            Miami Heat    98         0    19,6   
970        Denver Nuggets    97      Dallas Mavericks   113         0  20,302   

     New Date      %W %W as Home/Visitor Back to back  ...  Last N  \
0    22/10/19   0.073              0.176         0.00  ...    0.07   
1    22/10/19   0.049              0.180         0.00  ...    0.05   
2    23/10/19  -0.079             -0.013         0.00  ...   -0.08   
3    23/10/19   0.134              0.239         0.00  ...    0.13   
4    23/10/19   0.201              0.280         0.00  ...    0.20   
..        ...     ...                ...          ...  ...     ...   
966  10/03/20   0.330              0.412         0.00  ...    0.45   
967  11/03/20   0.286              0.652        -3.00  ...    0.33   
968  11/03/20  -0.005              0.143         1.00  ...    0.11   
969  11/03/20   0.297              0.507         1.00  ...    0.34   
970  11/03/20  -0.081             -0.036        -1.00  ...    0.00   

     Net Rating  OT last match  Rest/Injuries Visitor Stars  \
0         0.000           0.00                         0.00   
1         0.000           0.00                         0.00   
2         0.000           0.00                         0.00   
3         0.000           0.00                         1.00   
4         0.000           0.00                         0.00   
..          ...            ...                          ...   
966       8.400           0.00                         1.00   
967       5.230           0.00                         1.00   
968      -1.410           1.00                         0.00   
969      10.210          -1.00                         0.00   
970       2.650           0.00                         0.00   

    Rest/Injuries Home Stars Rest/Injuries Stars Win Streak Bet Visitor  \
0                       0.00                0.00       0.00        3.28   
1                       1.00                1.00       0.00        1.61   
2                       0.00                0.00       0.00        1.62   
3                       1.00                0.00       0.00        3.45   
4                       0.00                0.00       0.00        4.91   
..                       ...                 ...        ...         ...   
966                     0.00               -1.00       0.00        6.73   
967                     1.00                0.00       0.00        6.82   
968                     0.00                0.00       0.00        2.72   
969                     1.00                1.00       0.00        5.88   
970                     1.00                1.00       0.00        1.71   

    Bet Home   Bets  
0       1.36  -1.92  
1       2.39   0.78  
2       2.39   0.77  
3       1.33  -2.12  
4       1.20  -3.71  
..       ...    ...  
966     1.12  -5.61  
967     1.12  -5.70  
968     1.48  -1.24  
969     1.15  -4.73  
970     2.21   0.50  

[971 rows x 25 columns]

#### Calculate distance traveled for every team without playing as local for every match played as local we muliply the distance made per 0.5.

In [26]:
def traveling(data,dim,data_distance,dim_distance):
    travels={}
    for i in range(dim_distance):
        travels[str(data_distance['City Origin'][i])+' - '+str(data_distance['City Destiny'][i])]=data_distance['Distance (Km)'][i]
   
    kilometers={}
    last_match={}
    km_local=[]
    km_visitor=[]
    for i in range(dim):

        if data['Home'][i] not in kilometers:
            kilometers[data['Home'][i]]=0
            km_local=np.append(km_local,"{0:.2f}".format(float(kilometers[data['Home'][i]])))
        if data['Home'][i] not in last_match:
            last_match[data['Home'][i]]=data['Home'][i]
        if data['Visitor'][i] not in kilometers:
            kilometers[data['Visitor'][i]]=float(travels[str(data['Home'][i])+' - '+str(data['Visitor'][i])])
        if data['Visitor'][i] not in last_match:
            last_match[data['Visitor'][i]]=data['Home'][i]
            km_visitor=np.append(km_visitor,"{0:.2f}".format(float(kilometers[data['Visitor'][i]])))


        else:
            if  last_match[data['Home'][i]]==data['Home'][i]:
                kilometers[data['Visitor'][i]]=kilometers[data['Visitor'][i]]+float(travels[str(data['Home'][i])+' - '+str(last_match[data['Visitor'][i]])])
                kilometers[data['Home'][i]]=kilometers[data['Home'][i]]*0.5
                last_match[data['Visitor'][i]]=data['Home'][i]
                last_match[data['Home'][i]]=data['Home'][i]
                km_local=np.append(km_local,"{0:.2f}".format(float(kilometers[data['Home'][i]])))
                km_visitor=np.append(km_visitor,"{0:.2f}".format(float(kilometers[data['Visitor'][i]])))
            else:
                kilometers[data['Visitor'][i]]=kilometers[data['Visitor'][i]]+float(travels[str(data['Home'][i])+' - '+str(last_match[data['Visitor'][i]])])
                kilometers[data['Home'][i]]=kilometers[data['Home'][i]]+float(travels[str(data['Home'][i])+' - '+str(last_match[data['Home'][i]])])
                last_match[data['Visitor'][i]]=data['Home'][i]
                last_match[data['Home'][i]]=data['Home'][i]
                km_local=np.append(km_local,"{0:.2f}".format(float(kilometers[data['Home'][i]])))
                km_visitor=np.append(km_visitor,"{0:.2f}".format(float(kilometers[data['Visitor'][i]])))
       



    return km_local,km_visitor,last_match,kilometers


km_l,km_visit,last_m,kilo=traveling(data,len(data),data_distance,len(data_distance))

km=[]
for i in range(len(km_l)):
    km=np.append(km,"{0:.2f}".format(float(km_l[i])-float(km_visit[i])))

data['Km']=km
data


Visitor  PTSV                  Home  PTSH  Overtime Attend.  \
0    New Orleans Pelicans   122       Toronto Raptors   130         1  20,787   
1      Los Angeles Lakers   102  Los Angeles Clippers   112         0  19,068   
2           Chicago Bulls   125     Charlotte Hornets   126         0  15,424   
3         Detroit Pistons   119        Indiana Pacers   110         0  17,923   
4     Cleveland Cavaliers    85         Orlando Magic    94         0  18,846   
..                    ...   ...                   ...   ...       ...     ...   
966         Brooklyn Nets   104    Los Angeles Lakers   102         0  18,997   
967       Detroit Pistons   106    Philadelphia 76ers   124         0  20,172   
968       New York Knicks   136         Atlanta Hawks   131         1  15,393   
969     Charlotte Hornets   109            Miami Heat    98         0    19,6   
970        Denver Nuggets    97      Dallas Mavericks   113         0  20,302   

     New Date      %W %W as Home/Visitor Back to back  ...  Net Rating  \
0    22/10/19   0.073              0.176         0.00  ...       0.000   
1    22/10/19   0.049              0.180         0.00  ...       0.000   
2    23/10/19  -0.079             -0.013         0.00  ...       0.000   
3    23/10/19   0.134              0.239         0.00  ...       0.000   
4    23/10/19   0.201              0.280         0.00  ...       0.000   
..        ...     ...                ...          ...  ...         ...   
966  10/03/20   0.330              0.412         0.00  ...       8.400   
967  11/03/20   0.286              0.652        -3.00  ...       5.230   
968  11/03/20  -0.005              0.143         1.00  ...      -1.410   
969  11/03/20   0.297              0.507         1.00  ...      10.210   
970  11/03/20  -0.081             -0.036        -1.00  ...       2.650   

     OT last match  Rest/Injuries Visitor Stars  Rest/Injuries Home Stars  \
0             0.00                         0.00                      0.00   
1             0.00                         0.00                      1.00   
2             0.00                         0.00                      0.00   
3             0.00                         1.00                      1.00   
4             0.00                         0.00                      0.00   
..             ...                          ...                       ...   
966           0.00                         1.00                      0.00   
967           0.00                         1.00                      1.00   
968           1.00                         0.00                      0.00   
969          -1.00                         0.00                      1.00   
970           0.00                         0.00                      1.00   

    Rest/Injuries Stars Win Streak Bet Visitor Bet Home   Bets        Km  
0                  0.00       0.00        3.28     1.36  -1.92  -1789.00  
1                  1.00       0.00        1.61     2.39   0.78      0.00  
2                  0.00       0.00        1.62     2.39   0.77   -947.10  
3                  0.00       0.00        3.45     1.33  -2.12   -403.40  
4                  0.00       0.00        4.91     1.20  -3.71  -1436.00  
..                  ...        ...         ...      ...    ...       ...  
966               -1.00       0.00        6.73     1.12  -5.61  -1688.56  
967                0.00       0.00        6.82     1.12  -5.70   4296.74  
968                0.00       0.00        2.72     1.48  -1.24    229.58  
969                1.00       0.00        5.88     1.15  -4.73   1506.34  
970                1.00       0.00        1.71     2.21   0.50  -4856.11  

[971 rows x 26 columns]

#### Calculates if the win belongs to the visitor 0 or the local team 1

In [27]:
def winner(data,dim):
    winner=[]
    for i in range(dim):
        if int(data['PTSV'][i])>int(data['PTSH'][i]):
            winner=np.append(winner,0)
        else:
            winner=np.append(winner,1)
    return winner

win=winner(data,len(data))
#print(len(win))

data['Winner']=win
data

Visitor  PTSV                  Home  PTSH  Overtime Attend.  \
0    New Orleans Pelicans   122       Toronto Raptors   130         1  20,787   
1      Los Angeles Lakers   102  Los Angeles Clippers   112         0  19,068   
2           Chicago Bulls   125     Charlotte Hornets   126         0  15,424   
3         Detroit Pistons   119        Indiana Pacers   110         0  17,923   
4     Cleveland Cavaliers    85         Orlando Magic    94         0  18,846   
..                    ...   ...                   ...   ...       ...     ...   
966         Brooklyn Nets   104    Los Angeles Lakers   102         0  18,997   
967       Detroit Pistons   106    Philadelphia 76ers   124         0  20,172   
968       New York Knicks   136         Atlanta Hawks   131         1  15,393   
969     Charlotte Hornets   109            Miami Heat    98         0    19,6   
970        Denver Nuggets    97      Dallas Mavericks   113         0  20,302   

     New Date      %W %W as Home/Visitor Back to back  ...  OT last match  \
0    22/10/19   0.073              0.176         0.00  ...           0.00   
1    22/10/19   0.049              0.180         0.00  ...           0.00   
2    23/10/19  -0.079             -0.013         0.00  ...           0.00   
3    23/10/19   0.134              0.239         0.00  ...           0.00   
4    23/10/19   0.201              0.280         0.00  ...           0.00   
..        ...     ...                ...          ...  ...            ...   
966  10/03/20   0.330              0.412         0.00  ...           0.00   
967  11/03/20   0.286              0.652        -3.00  ...           0.00   
968  11/03/20  -0.005              0.143         1.00  ...           1.00   
969  11/03/20   0.297              0.507         1.00  ...          -1.00   
970  11/03/20  -0.081             -0.036        -1.00  ...           0.00   

     Rest/Injuries Visitor Stars  Rest/Injuries Home Stars  \
0                           0.00                      0.00   
1                           0.00                      1.00   
2                           0.00                      0.00   
3                           1.00                      1.00   
4                           0.00                      0.00   
..                           ...                       ...   
966                         1.00                      0.00   
967                         1.00                      1.00   
968                         0.00                      0.00   
969                         0.00                      1.00   
970                         0.00                      1.00   

     Rest/Injuries Stars Win Streak Bet Visitor Bet Home   Bets        Km  \
0                   0.00       0.00        3.28     1.36  -1.92  -1789.00   
1                   1.00       0.00        1.61     2.39   0.78      0.00   
2                   0.00       0.00        1.62     2.39   0.77   -947.10   
3                   0.00       0.00        3.45     1.33  -2.12   -403.40   
4                   0.00       0.00        4.91     1.20  -3.71  -1436.00   
..                   ...        ...         ...      ...    ...       ...   
966                -1.00       0.00        6.73     1.12  -5.61  -1688.56   
967                 0.00       0.00        6.82     1.12  -5.70   4296.74   
968                 0.00       0.00        2.72     1.48  -1.24    229.58   
969                 1.00       0.00        5.88     1.15  -4.73   1506.34   
970                 1.00       0.00        1.71     2.21   0.50  -4856.11   

    Winner  
0      1.0  
1      1.0  
2      1.0  
3      0.0  
4      1.0  
..     ...  
966    0.0  
967    1.0  
968    0.0  
969    0.0  
970    1.0  

[971 rows x 27 columns]

In [28]:
#we store all the data collected and processed to the following .csv file
data.to_csv('NBA_analysis_19_20.csv', index=False)